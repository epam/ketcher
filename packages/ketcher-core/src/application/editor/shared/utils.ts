/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { Struct, Vec2 } from 'domain/entities'
import { inRange } from 'lodash'
import { BondAtoms } from 'application/editor/shared/utils.types'

let FRAC = Math.PI / 12 // '15ยบ'

function setFracAngle(angle: number): void {
  FRAC = (Math.PI / 180) * angle
}

function calcAngle(pos0: Vec2, pos1: Vec2): number {
  const v = Vec2.diff(pos1, pos0)
  return Math.atan2(v.y, v.x)
}

function fracAngle(angle, angle2): number {
  if (angle2) angle = calcAngle(angle, angle2)
  return Math.round(angle / FRAC) * FRAC
}

function calcNewAtomPos(pos0: Vec2, pos1: Vec2, ctrlKey: boolean): Vec2 {
  const vector = new Vec2(1, 0).rotate(
    ctrlKey ? calcAngle(pos0, pos1) : fracAngle(pos0, pos1)
  )
  vector.add_(pos0) // eslint-disable-line no-underscore-dangle
  return vector
}

function degrees(angle: number): number {
  let degree = Math.round((angle / Math.PI) * 180)
  if (degree > 180) degree -= 360
  else if (degree <= -180) degree += 360
  return degree
}

const BONDS_MERGE_ANGLE = 10 // 'ยบ'
const BONDS_MERGE_SCALE = 0.2

function mergeBondsParams(
  struct1: Struct,
  bond1: BondAtoms,
  struct2: Struct,
  bond2: BondAtoms
) {
  // TODO find a solution to remove not-null assertion from result atoms below
  // https://github.com/epam/ketcher/issues/2652
  const begin1 = struct1.atoms.get(bond1.begin)!
  const begin2 = struct2.atoms.get(bond2.begin)!
  const end1 = struct1.atoms.get(bond1.end)!
  const end2 = struct2.atoms.get(bond2.end)!

  const angle = calcAngle(begin1.pp, end1.pp) - calcAngle(begin2.pp, end2.pp)
  const mergeAngle = Math.abs(degrees(angle) % 180)

  const scale = Vec2.dist(begin1.pp, end1.pp) / Vec2.dist(begin2.pp, end2.pp)

  const merged =
    !inRange(mergeAngle, BONDS_MERGE_ANGLE, 180 - BONDS_MERGE_ANGLE) &&
    inRange(scale, 1 - BONDS_MERGE_SCALE, 1 + BONDS_MERGE_SCALE)

  return { merged, angle, scale, cross: Math.abs(degrees(angle)) > 90 }
}

export default {
  calcAngle,
  fracAngle,
  calcNewAtomPos,
  degrees,
  setFracAngle,
  mergeBondsParams
}
