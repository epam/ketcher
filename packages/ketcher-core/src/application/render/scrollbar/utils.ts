import type { RenderOptions } from '../render.types';

const SCALE_FACTOR = 0.5;

/**
 * To make scrollbar provide better UX, this function can help to achieve
 * moving scrollbar `x` offsets leads to moving viewBox `2 * x` offsets
 * */
export const getUserFriendlyScrollOffset = (offset: number) => {
  return offset * SCALE_FACTOR;
};

export const getUserFriendlyViewBoxDelta = (delta: number) => {
  return delta / SCALE_FACTOR;
};

export const getZoomedValue = (value: number, renderOptions: RenderOptions) => {
  return value / renderOptions.zoom;
};
