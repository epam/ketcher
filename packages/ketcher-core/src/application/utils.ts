import { Struct, Vec2 } from 'domain/entities';
import { DrawingEntitiesManager } from 'domain/entities/DrawingEntitiesManager';
import {
  FormatterFactory,
  SupportedFormat,
  identifyStructFormat,
} from './formatters';
import { Ketcher } from './ketcher';
import { ChemicalMimeType, StructService } from 'domain/services';
import { Coordinates, CoreEditor, EditorHistory } from './editor/internal';
import { KetSerializer } from 'domain/serializers';
import assert from 'assert';

class KetcherProvider {
  private ketcherInstance: Ketcher | undefined;

  setKetcherInstance(ketcherInstance: Ketcher) {
    this.ketcherInstance = ketcherInstance;
  }

  getKetcher() {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.ketcherInstance!;
  }
}

const ketcherProvider = new KetcherProvider();

export { ketcherProvider };

export function getStructure(
  structureFormat = SupportedFormat.rxn,
  formatterFactory: FormatterFactory,
  struct: Struct,
  drawingEntitiesManager?: DrawingEntitiesManager,
): Promise<string> {
  const formatter = formatterFactory.create(structureFormat);
  return formatter.getStructureFromStructAsync(struct, drawingEntitiesManager);
}

export async function prepareStructToRender(
  structStr: string,
  structService: StructService,
  ketcherInstance: Ketcher,
): Promise<Struct> {
  const struct: Struct = await parseStruct(
    structStr,
    structService,
    ketcherInstance,
  );
  struct.initHalfBonds();
  struct.initNeighbors();
  struct.setImplicitHydrogen();
  struct.markFragments();

  return struct;
}

export function parseStruct(
  structStr: string,
  structService: StructService,
  ketcherInstance: Ketcher,
) {
  const format = identifyStructFormat(structStr);
  const factory = new FormatterFactory(structService);
  const options = ketcherInstance.editor.options();

  const service = factory.create(format, {
    'dearomatize-on-load': options['dearomatize-on-load'],
    'ignore-no-chiral-flag': options.ignoreChiralFlag,
  });
  return service.getStructureFromStringAsync(structStr);
}

export function deleteAllEntitiesOnCanvas() {
  const editor = CoreEditor.provideEditorInstance();
  const modelChanges = editor.drawingEntitiesManager.deleteAllEntities();
  editor.renderersContainer.update(modelChanges);
}

export async function parseAndAddMacromoleculesOnCanvas(
  struct: string,
  structService: StructService,
) {
  const editor = CoreEditor.provideEditorInstance();
  const ketSerializer = new KetSerializer();
  const format = identifyStructFormat(struct);
  let ketStruct = struct;
  if (format !== SupportedFormat.ket) {
    ketStruct = (
      await structService.convert({
        struct,
        output_format: ChemicalMimeType.KET,
      })
    ).struct;
  }

  const deserialisedKet = ketSerializer.deserializeToDrawingEntities(ketStruct);
  assert(deserialisedKet);
  const modelChanges = deserialisedKet.drawingEntitiesManager.mergeInto(
    editor.drawingEntitiesManager,
  );

  new EditorHistory(editor).update(modelChanges);
  editor.renderersContainer.update(modelChanges);
}

export function getCurrentCenterPointOfCanvas() {
  const editor = CoreEditor.provideEditorInstance();
  const originalCenterPointOfCanvas = new Vec2(
    editor.canvasOffset.width / 2,
    editor.canvasOffset.height / 2,
  );
  return Coordinates.viewToCanvas(originalCenterPointOfCanvas);
}
