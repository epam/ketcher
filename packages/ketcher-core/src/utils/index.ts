export * from './Box2Abs'
export * from './Pile'
export * from './Pool'
export * from './scale'
export * from './Vec2'
