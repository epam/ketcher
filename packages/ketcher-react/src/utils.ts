/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

/**
 * Determines the appropriate HTML element for fullscreen mode based on the current DOM structure.
 *
 * Priority order:
 * 1. Element with [data-ketcher-fullscreen-container] attribute (explicit user customization)
 * 2. Parent of [data-ketcher-editor] element:
 *    - If parent is #root -> return #root (standalone mode)
 *    - If parent is custom container -> return parent (iframe/embedded mode)
 * 3. Fallback to #root or documentElement
 *
 * @returns {HTMLElement} The element to be used for fullscreen mode
 */
export const getFullscreenElement = (): HTMLElement => {
  const ketcherRoot = document.querySelector(
    '.Ketcher-root, .Ketcher-polymer-editor-root',
  ) as HTMLElement;

  if (ketcherRoot) {
    const markedContainer = ketcherRoot.closest(
      '[data-ketcher-fullscreen-container]',
    );
    if (markedContainer) {
      return markedContainer as HTMLElement;
    }

    const editorRoot = ketcherRoot.closest('[data-ketcher-editor]');
    if (editorRoot) {
      const editorParent = editorRoot.parentElement;

      if (
        editorParent &&
        editorParent !== document.body &&
        editorParent !== document.documentElement
      ) {
        if (editorParent.id === 'root') {
          return editorParent;
        }
        return editorParent;
      }

      const rootElement = editorRoot.closest('#root');
      if (rootElement) {
        return rootElement as HTMLElement;
      }
    }

    const rootById = document.getElementById('root');
    if (rootById) {
      return rootById;
    }
  }

  return document.getElementById('root') || document.documentElement;
};
