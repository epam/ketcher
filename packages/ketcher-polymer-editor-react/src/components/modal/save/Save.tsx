/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { useEffect, useState } from 'react'
import styled from '@emotion/styled'

import { Modal } from 'components/shared/modal'
import { DropDown } from 'components/shared/dropDown'
import { Option } from 'components/shared/dropDown/dropDown'
import { TextField } from 'components/shared/textEditor'
import { TextInputField } from 'components/shared/textInputField'
import { SaveButton } from 'components/modal/save/saveButton'
import { getPropertiesByFormat, SupportedFormats } from 'helpers/formats'
import { ActionButton } from 'components/shared/actionButton'
import { Icon } from 'ketcher-react'

interface Props {
  onClose: () => void
  isModalOpen: boolean
}

const options: Array<Option> = [
  { id: 'mol', label: 'MDL Molfile V3000' },
  { id: 'helm', label: 'HELM' }
]

const Form = styled.form({
  display: 'flex',
  flexDirection: 'column',
  height: '100%'
})

const Row = styled.div({
  display: 'flex',
  justifyContent: 'space-between',
  marginBottom: '16px'
})

const Label = styled.label(({ theme }) => ({
  marginRight: '8px',
  color: theme.ketcher.color.text.secondary
}))

const StyledDropdown = styled(DropDown)(({ theme }) => ({
  '& .MuiOutlinedInput-root': {
    border: 'none',
    backgroundColor: theme.ketcher.color.background.primary,
    color: theme.ketcher.color.text.primary,
    fontFamily: theme.ketcher.font.family.inter
  }
}))

const stylesForExpanded = {
  border: 'none'
}

const StyledModal = styled(Modal)({
  '& .MuiPaper-root': {
    width: '520px',
    height: '358px'
  },

  '& .MuiDialogContent-root': {
    overflow: 'hidden'
  }
})

const ErrorsButton = styled(ActionButton)(({ theme }) => ({
  backgroundColor: theme.ketcher.color.background.canvas,
  color: theme.ketcher.color.text.error,

  '&:hover': {
    backgroundColor: 'initial'
  }
}))

const structExample = {
  mol: `
    -INDIGO-02102212382D
  
    0  0  0  0  0  0  0  0  0  0  0 V3000
  M  V30 BEGIN CTAB
  M  V30 COUNTS 11 10 0 0 0
  M  V30 BEGIN ATOM
  M  V30 1 C 2.75 -5.225 0.0 0
  M  V30 2 C 3.61603 -5.725 0.0 0
  M  V30 3 C 4.48205 -5.225 0.0 0
  M  V30 4 C 5.34808 -5.725 0.0 0
  M  V30 5 C 6.2141 -5.225 0.0 0
  M  V30 6 C 7.08013 -5.725 0.0 0
  M  V30 7 C 7.94615 -5.225 0.0 0
  M  V30 8 C 8.81218 -5.725 0.0 0
  M  V30 9 C 9.6782 -5.225 0.0 0
  M  V30 10 C 10.5442 -5.725 0.0 0
  M  V30 11 C 11.4103 -5.225 0.0 0
  M  V30 END ATOM
  M  V30 BEGIN BOND
  M  V30 1 1 1 2
  M  V30 2 1 2 3
  M  V30 3 1 3 4
  M  V30 4 1 4 5
  M  V30 5 1 5 6
  M  V30 6 1 6 7
  M  V30 7 1 7 8
  M  V30 8 1 8 9
  M  V30 9 1 9 10
  M  V30 10 1 10 11
  M  V30 END BOND
  M  V30 END CTAB
  M  END
  `,
  helm: `PEPTIDE1{A.C.D.F.G.H.K.A.C.D}$PEPTIDE1,PEPTIDE1,3:R3-7:R3`
} // TODO remove when canvas and get struct method are ready

export const Save = ({ onClose, isModalOpen }: Props): JSX.Element => {
  const [currentFileFormat, setCurrentFileFormat] =
    useState<SupportedFormats>('mol')
  const [currentFileName, setCurrentFileName] = useState('ketcher')
  const [struct, setStruct] = useState('')
  const [errors, setErrors] = useState('')

  const handleSelectChange = (value) => {
    setCurrentFileFormat(value)
  }

  const handleInputChange = (value) => {
    setCurrentFileName(value)
  }

  const handleSave = () => {
    console.log('Saved', structExample)
  }

  const handleErrorsClick = () => {
    console.log('errors...')
  }

  useEffect(() => {
    console.log('Getting and setting struct here...') // get / convert struct and errors
    setStruct(structExample[currentFileFormat])

    if (currentFileFormat === 'mol') {
      // just an example
      setErrors('some error')
    } else {
      setErrors('')
    }
  }, [currentFileFormat])

  return (
    <StyledModal title="save structure" isOpen={isModalOpen} onClose={onClose}>
      <Modal.Content>
        <Form onSubmit={handleSave} id="save">
          <Row>
            <div>
              <TextInputField
                value={currentFileName}
                id="filename"
                onChange={handleInputChange}
                label="File name:"
              />
            </div>
            <div style={{ lineHeight: '24px' }}>
              <Label htmlFor="fileformat">File format:</Label>
              <StyledDropdown
                options={options}
                currentSelection={currentFileFormat}
                selectionHandler={handleSelectChange}
                customStylesForExpanded={stylesForExpanded}
              />
            </div>
          </Row>
          <TextField struct={struct} readonly selectOnInit />
        </Form>
      </Modal.Content>

      <Modal.Footer>
        {errors && (
          <ErrorsButton label="Errors" clickHandler={handleErrorsClick}>
            <Icon name="error" />
            Errors
          </ErrorsButton>
        )}

        <SaveButton
          label="Save as file"
          data={struct}
          type={getPropertiesByFormat(currentFileFormat).mime}
          onSave={handleSave}
          filename={
            currentFileName +
            getPropertiesByFormat(currentFileFormat).extensions[0]
          }
          disabled={!currentFileName}
        />
      </Modal.Footer>
    </StyledModal>
  )
}
