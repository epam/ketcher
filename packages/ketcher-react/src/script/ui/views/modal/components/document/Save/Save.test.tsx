import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Save from '.';
import { renderWithMockContext } from './Save.test.utils';

describe('Save Dialog should be rendered correctly', () => {
  it('should render opened file format dropdown when the closed dropdown is clicked', async () => {
    const view = renderWithMockContext(<Save />);

    await userEvent.click(screen.getByText('MDL Molfile V2000'));
    await screen.findByText('MDL Molfile V3000');

    expect(view).toMatchSnapshot();
  });
});
