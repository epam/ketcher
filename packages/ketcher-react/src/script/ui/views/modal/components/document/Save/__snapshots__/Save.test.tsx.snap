// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save Dialog should be rendered correctly should render opened file format dropdown when the closed dropdown is clicked 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div
        class="dialog dialog"
        role="dialog"
        tabindex="-1"
      >
        <header
          class="header withDivider"
        >
          <span>
            Save Structure
          </span>
          <div
            class="btnContainer"
          >
            <button
              class="buttonTop"
            >
              <icon-close.svg
                classname="closeButton"
              />
            </button>
          </div>
        </header>
        <div
          class="body"
        >
          <div
            class="formContainer"
          >
            <form>
              <label
                class=""
              >
                <span>
                  File name:
                </span>
                <span
                  class="inputWrapper"
                >
                  <input
                    class="input genericInput"
                    name="filename"
                    type="text"
                    value=""
                  />
                </span>
              </label>
              <label
                class=""
              >
                <span>
                  File format:
                </span>
                <span
                  class="inputWrapper"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused selectContainer css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                    data-testid="file-format-dropdown"
                  >
                    <div
                      aria-controls=":r0:"
                      aria-expanded="true"
                      aria-haspopup="listbox"
                      class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                      role="combobox"
                      tabindex="0"
                    >
                      MDL Molfile V2000
                    </div>
                    <input
                      aria-hidden="true"
                      aria-invalid="false"
                      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                      tabindex="-1"
                      value="mol"
                    />
                    <icon-chevron.svg
                      classname="MuiSelect-icon MuiSelect-iconOutlined MuiSelect-iconOpen css-10q54uo-MuiSelect-icon chevronIcon"
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-ihdtdm"
                      >
                        <span
                          class="notranslate"
                        >
                          â€‹
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </span>
              </label>
            </form>
            <div>
              <ul
                class="tabs"
                tabindex="0"
              >
                <li
                  class="tabs"
                >
                  <a
                    class="active"
                  >
                    Preview
                  </a>
                </li>
              </ul>
              <div>
                <div
                  class="previewBackground"
                >
                  <textarea
                    class="previewArea"
                    readonly=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="footer"
        >
          <button
            class="saveTmpl"
            disabled=""
          >
            Save to Templates
          </button>
          <button
            class="cancel"
            mode="onCancel"
            type="button"
          >
            Cancel
          </button>
          <button
            class="ok"
            disabled=""
          >
            Save
          </button>
        </footer>
      </div>
    </div>
    <div
      class="MuiPopover-root MuiMenu-root dropdownList MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
      id="menu-"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
        style="opacity: 1; transform: scale(1, 1); min-width: 0; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
          id=":r0:"
          role="listbox"
          tabindex="-1"
        >
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="ket"
            role="option"
            tabindex="-1"
          >
            Ket Format
          </li>
          <li
            aria-selected="true"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-selected MuiMenuItem-root MuiMenuItem-gutters Mui-selected css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="mol"
            role="option"
            tabindex="0"
          >
            MDL Molfile V2000
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="molV3000"
            role="option"
            tabindex="-1"
          >
            MDL Molfile V3000
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="sdf"
            role="option"
            tabindex="-1"
          >
            SDF V2000
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="sdfV3000"
            role="option"
            tabindex="-1"
          >
            SDF V3000
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="smarts"
            role="option"
            tabindex="-1"
          >
            Daylight SMARTS
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="smiles"
            role="option"
            tabindex="-1"
          >
            Daylight SMILES
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="smilesExt"
            role="option"
            tabindex="-1"
          >
            Extended SMILES
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="cml"
            role="option"
            tabindex="-1"
          >
            CML
          </li>
          <hr
            aria-selected="false"
            class="MuiDivider-root MuiDivider-fullWidth listDivider css-9mgopn-MuiDivider-root"
            role="option"
          />
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="inChI"
            role="option"
            tabindex="-1"
          >
            InChI
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="inChIAuxInfo"
            role="option"
            tabindex="-1"
          >
            InChI AuxInfo
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="inChIKey"
            role="option"
            tabindex="-1"
          >
            InChIKey
          </li>
          <hr
            aria-selected="false"
            class="MuiDivider-root MuiDivider-fullWidth listDivider css-9mgopn-MuiDivider-root"
            role="option"
          />
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="svg"
            role="option"
            tabindex="-1"
          >
            SVG Document
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="png"
            role="option"
            tabindex="-1"
          >
            PNG Image
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="cdxml"
            role="option"
            tabindex="-1"
          >
            CDXML
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="cdx"
            role="option"
            tabindex="-1"
          >
            Base64 CDX
          </li>
          <li
            aria-selected="false"
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            data-value="binaryCdx"
            role="option"
            tabindex="-1"
          >
            CDX
          </li>
        </ul>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      class="dialog dialog"
      role="dialog"
      tabindex="-1"
    >
      <header
        class="header withDivider"
      >
        <span>
          Save Structure
        </span>
        <div
          class="btnContainer"
        >
          <button
            class="buttonTop"
          >
            <icon-close.svg
              classname="closeButton"
            />
          </button>
        </div>
      </header>
      <div
        class="body"
      >
        <div
          class="formContainer"
        >
          <form>
            <label
              class=""
            >
              <span>
                File name:
              </span>
              <span
                class="inputWrapper"
              >
                <input
                  class="input genericInput"
                  name="filename"
                  type="text"
                  value=""
                />
              </span>
            </label>
            <label
              class=""
            >
              <span>
                File format:
              </span>
              <span
                class="inputWrapper"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused selectContainer css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                  data-testid="file-format-dropdown"
                >
                  <div
                    aria-controls=":r0:"
                    aria-expanded="true"
                    aria-haspopup="listbox"
                    class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                    role="combobox"
                    tabindex="0"
                  >
                    MDL Molfile V2000
                  </div>
                  <input
                    aria-hidden="true"
                    aria-invalid="false"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    tabindex="-1"
                    value="mol"
                  />
                  <icon-chevron.svg
                    classname="MuiSelect-icon MuiSelect-iconOutlined MuiSelect-iconOpen css-10q54uo-MuiSelect-icon chevronIcon"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        â€‹
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </span>
            </label>
          </form>
          <div>
            <ul
              class="tabs"
              tabindex="0"
            >
              <li
                class="tabs"
              >
                <a
                  class="active"
                >
                  Preview
                </a>
              </li>
            </ul>
            <div>
              <div
                class="previewBackground"
              >
                <textarea
                  class="previewArea"
                  readonly=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <footer
        class="footer"
      >
        <button
          class="saveTmpl"
          disabled=""
        >
          Save to Templates
        </button>
        <button
          class="cancel"
          mode="onCancel"
          type="button"
        >
          Cancel
        </button>
        <button
          class="ok"
          disabled=""
        >
          Save
        </button>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
